### ISSUE TITLE
Win Condition Suggestion (BDRA Type)

### ISSUE THREAD

**Issue by CerberusVelvet** [RotorHazard/RotorHazard#428]
An enhancement suggestion for RH. To help make it usable to run race practice sessions in the UK to mirror the race config in official uk qualifiers. Maybe even move towards making RH usable to run a UK qualifier!

In the UK our official British Drone Racing Association races are all run on LiveTime. All UK qualifying events are run the same way.

I'll explain how it works.....

The format is a two or three minute race (usually). Once that timer elapses there is a 'finish your lap' timer that is usually thirty seconds. Which allows each pilot the chance to complete their current lap once the timer has elapsed. If they can do so within the 'finish your lap' time allowed.

The winner of the heat is the pilot who completes the most laps in the fastest time.

For example. Given a two minute race with a thirty second 'finish your laps' timer.
At two minutes the timer announces "Time elapsed, finish your laps" (or something user selected)
After the 'finish your laps' time the timer announces "Race is completed, winner is XX. Please land your quads" (or something user selected)

Pilot A completes three laps in the two minutes. When the two minutes expires Pilot A is partway into their fourth lap. They complete that lap within the allocated thirty seconds. When crossing the timer the announcement is "Pilot A, Done". Pilot then lands in the designated landing zone of the event. Finishing with four laps in two minutes twenty seconds.

Pilot B completes four laps in the two minutes. When the two minutes expires Pilot B is on their fifth lap. However they crash and do not complete their fifth lap within the thirty seconds. Finishing with four laps in one minute fifty seconds.

Pilot C completes two laps in the two minutes. When the two minutes expires Pilot C is partway into their third lap. However they don't make it around in time and they land the quad in the designated landing zone when they hear the announcement that the race is complete. Finishing with two laps in one minute thirty seconds.

Once the race timer is complete and after the 'finish your laps' timer. All quads should have landed, crashed out, or be moving towards the designated landing zone. The race has ended and the winner is announced, with times "Race is complete. The winner is Pilot B, Four laps in One minute fifty" The race then auto stops on RH. 

The winner of this race is Pilot B. Most laps in the shortest time. Followed by Pilot A in second and Pilot C in third.

Once the race director is happy that all quads are landed and the track is clear the timer announces "Track is clear, collect your quads" (or similar as user selected). This would require a 'track clear' button on the UI? Or could be triggered by saving the laps?This allows the RD to safely control the event, giving a clear signal that they track is clear for pilots to find the remains of their quad buried in a gate.

Proposed changes enhancements to RH
A win condition of "Fastest Laps with Finish Your Laps" in race format settings.
A configurable 'Finish Your Lap' time in race format settings that is active when above format is selected
A set of configurable callouts that are auto triggered by race states. 'Callout - timer complete' / 'callout - race complete' / 'callout - track clear'. (These would also be used by other race formats too) Making each race ending callout configurable by the RD.
This may require the addition of some new %variable%s such as %racewinner% %winninglaps% %winningtime% for example?

Hope that all makes sense!

**Comment by HazardCreative** [comment]
Thanks for the very clear writeup! This should be split into a few separate issues, though, so they can be tracked separately as they are resolved.

You'll be interested in the changes currently in master that we are just about to publish to beta. These improve win conditions and intelligently determine the winner when possible in each type of race.

Announcing that an individual pilot is done after completing their last lap is another separate issue that could be filed.

**Comment by CerberusVelvet** [comment]
Good news. Please do split it up into however it makes most sense for you guys! Glad it made some sense ;)

I've just downloaded the latest master and installed to my timer today. Looking forward to giving the latest win conditions a test!

**Comment by HazardCreative** [comment]
Issues have been broken out into individual pieces. Closing this compound issue.
